bsp/Src/BSP_ADC.c

```diff
@@ -0,0 +1,163 @@
+#include "BSP_ADC.h"
+
+ADC_HandleTypeDef hadc;
+
+double ADC_RAW_TO_VOLTAGE;  // conversion factor for readings
```

Comment:
This conversion can be made a bit more accurate by utilizing fixed point. Instead of storing a double here to represent the decimal, you can store both the VCC and bitNum and do the multiplication and division in the interrupt itself, to preserve some accuracy especially at higher values.
Here's a few resources to make this change:
https://stackoverflow.com/a/67234824
https://www.tonmeister.ca/wordpress/2021/07/19/fixed-point-vs-floating-point/

Reply:
Additionally, some macro magic can probably be used to make defining the bitNum and VCC for each ADC channel a compile time step rather than a runtime step. To sort of understand how this would work, take a look at Nathaniel's PR (https://github.com/lhr-solar/Embedded-Sharepoint/pull/76/files), specifically at how CAN.c and can1_recv_entries.h are used in conjunction. The macro stuff for this wouldn't need to be as complex, but understanding how that works wil give you a good starting point for understanding how to implement something that would work in this context.
This stuff can be a separate PR if you wish (make an issue ticket for it if this is the case). For right now, I'd say changing this to fixed point could be enough.

Reply:
Finally, this can be static as well. static just means it's a variable that's file-local, not that it necessarily is a constant (const is the keyword used for constants)